File: src/types/file.ts
import type { Frontmatter } from "./note";

export interface FileEntry {
  path: string;
  name: string;
  isDir: boolean;
  children?: FileEntry[];
}

export interface FileMetadata {
  path: string;
  sizeBytes: number;
  modifiedAt: string;
  createdAt: string;
  fileType: FileType;
  frontmatter: Frontmatter | null;
  tags: string[];
  outgoingLinks: string[];
  gitStatus: GitFileStatus | null;
}

export type FileType =
  | { type: "markdown" }
  | { type: "code"; language: string }
  | { type: "image" }
  | { type: "binary" }
  | { type: "other" };

export type GitFileStatus =
  | "unmodified"
  | "modified"
  | "added"
  | "deleted"
  | "renamed"
  | "untracked"
  | "conflicted";

// Re-export Frontmatter for convenience
export type { Frontmatter };

File: src-tauri/src/models/file.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileEntry {
    pub path: String,
    pub name: String,
    pub is_dir: bool,
    pub children: Option<Vec<FileEntry>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileMetadata {
    pub path: String,
    pub size_bytes: u64,
    pub modified_at: String,
    pub created_at: String,
    pub file_type: FileType,
    pub frontmatter: Option<serde_json::Value>,
    pub tags: Vec<String>,
    pub outgoing_links: Vec<String>,
    pub git_status: Option<GitFileStatus>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum FileType {
    Markdown,
    Code { language: String },
    Image,
    Binary,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GitFileStatus {
    Unmodified,
    Modified,
    Added,
    Deleted,
    Renamed,
    Untracked,
    Conflicted,
}
